onyo/lib/onyo.py

```diff
@@ -906,6 +930,227 @@ def valid_name(self, asset: Path) -> bool:
 
         return True
 
+    #
+    # SET
+    #
+    def set(self, paths: Iterable[Union[Path, str]], values: dict, yes: bool,
+            recursive: bool, rename: bool, depth: Union[int]) -> str:
+        """
+        Set values for a list of assets (or directories), or rename assets
+        through updating their name fields.
+
+        Flags enable to set values in directories optionally recursively, and/or
+        limit the depth of recursion.
+        """
+        assets_to_set = self._set_sanitize(paths, recursive, depth)
+
+        content_values = dict((field, values[field]) for field in values.keys() if field not in ["type", "make", "model", "serial"])
+        name_values = dict((field, values[field]) for field in values.keys() if field in ["type", "make", "model", "serial"])
+
+        if name_values and not rename:
+            log.error("Can't change pseudo keys without --rename.")
+            raise ValueError("Can't change pseudo keys without --rename.")
+
+        if content_values:
+            for asset in assets_to_set:
+                contents = self._read_asset(asset)
+                contents.update(content_values)
+                self._write_asset(asset, contents)
+                self.add(asset)
+
+        if name_values:
+            try:
+                self._update_names(assets_to_set, name_values)
+                self.clear_caches()
+            except ValueError as e:
+                self.clear_caches()
+                if self.files_staged:
+                    self._git(['restore', '--source=HEAD', '--staged', '--worktree'] +
+                              [str(file) for file in self.files_staged])
+                # reset renaming needs double-restoring
+                if self.files_staged:
+                    self._git(['restore', '--source=HEAD', '--staged', '--worktree'] +
+                              [str(file) for file in self.files_staged])
+                raise ValueError(e)
+
+        return self._diff_changes()
+
+    def _select_assets_from_directory(self, directory: Path, depth: Union[int, None],
```

Comment:
The only consumer of this is `_set_sanitize()`, yet there is a significant amount of overlap of functionality and checks between the two functions.

I have not thought about this at length, but my impulse is to consolidate these. Perhaps getting a list of all assets/directories first, and /then/ checking them?

Reply:
This function might also be re-usable for later commands like `onyo get`, and depending on how we implement `onyo ls`.

Alternatively, depending on how we solve https://github.com/psyinfra/onyo/issues/259 this might also have an impact of the code of both functions `_set_sanitize()` and `_select_assets_from_directory()`, since the functionality would be similar.

This is not to say it should never be changed again, but I think it might be more valueable as a separate part for now and be changed then together with the other issue.